aws-codeguruprofiler-profilinggroup/src/main/java/software/amazon/codeguruprofiler/profilinggroup/CreateHandler.java

```diff
@@ -17,42 +21,90 @@
+
+        Optional<List<String>> principals = principalsForAgentPermissionsFrom(model);
+        if (principals.isPresent()) {
+            putAgentPermissions(proxy, pgName, principals.get());
+            logger.log(format("%s [%s] for accountId [%s] has been successfully updated with agent permissions!",
+                    ResourceModel.TYPE_NAME, pgName, awsAccountId));
+        }
+
+        return ProgressEvent.defaultSuccessHandler(model);
+    }
 
-            logger.log(String.format("%s [%s] for accountId [%s] has been successfully created!", ResourceModel.TYPE_NAME, model.getProfilingGroupName(), awsAccountId));
+    private void putAgentPermissions(final AmazonWebServicesClientProxy proxy, final String pgName, final List<String> principals) {
+        PutPermissionRequest putPermissionRequest = PutPermissionRequest.builder()
+                .profilingGroupName(pgName)
+                .actionGroup(ActionGroup.AGENT_PERMISSIONS)
+                .principals(principals)
+                .build();
+        try {
+            proxy.injectCredentialsAndInvokeV2(putPermissionRequest, profilerClient::putPermission);
+        } catch (CodeGuruProfilerException e) {
+            try {
+                DeleteProfilingGroupRequest deletePgRequest = DeleteProfilingGroupRequest.builder().profilingGroupName(pgName).build();
+                proxy.injectCredentialsAndInvokeV2(deletePgRequest, profilerClient::deleteProfilingGroup);
+            } catch (Throwable deleteException) {
+                e.addSuppressed(deleteException);
+            } finally {
+                wrapException(e);
+            }
+        }
+    }
 
-            return ProgressEvent.defaultSuccessHandler(model);
-        } catch (ConflictException e) {
+    private static void wrapException(final CodeGuruProfilerException e) {
+        if (e instanceof ConflictException) {
             throw new CfnAlreadyExistsException(e);
-        } catch (InternalServerException e) {
+        }
+        if (e instanceof InternalServerException) {
             throw new CfnServiceInternalErrorException(e);
-        } catch (ServiceQuotaExceededException e) {
+        }
+        if (e instanceof ServiceQuotaExceededException) {
             throw new CfnServiceLimitExceededException(e);
-        } catch (ThrottlingException e) {
+        }
+        if (e instanceof ThrottlingException) {
             throw new CfnThrottlingException(e);
-        } catch (ValidationException e) {
+        }
+        if (e instanceof ValidationException) {
             throw new CfnInvalidRequestException(ResourceModel.TYPE_NAME + e.getMessage(), e);
         }
+        throw e;
+    }
+
+    private static Optional<List<String>> principalsForAgentPermissionsFrom(final ResourceModel model) {
+        if (model.getAgentPermissions() == null) {
+            return Optional.empty();
+        }
+        return Optional.ofNullable(model.getAgentPermissions().getPrincipals());
```

Comment:
No, principals is required through the schema. Submitting with an empty array causes _invalid request provided: AWS::CodeGuruProfiler::ProfilingGroup1 validation error detected: Value '[]' at 'principals' failed to satisfy constraint: Member must have length greater than or equal to 1_.

So, it can not be null, so I will keep only ```Optional.of(...)```, thanks.